LTI Deep-Linking

kristofkeppens opened the pull request:
This PR adds a new LTI tool, deeplink. Which enables users to easily add opencast content to their LMS.
This implementation was originally done by @ppettit , but when we had the LTI module rewritten to react this was no longer compatible. This fixes this by implementing this feature within the new LTI react module.

We tested this against our own LMS ( brightspace ), but behaviour might vary with other LMS'es.

kristofkeppens commented:
We'll have a look at the style errors, we're not getting those locally for some reason ...

mtneug committed:
educast: Don't force inclusion of referred elements in snapshot

When a referred element should be excluded, but the reference has
properties Opencast forces the element to stay in the snapshot. This
changes the behaviour to the same as when no properties are set, i.e.
the refered element will not be included in the snapshot when a user
requests this.

Fixes #2627

wsmirnow commented:
@kristofkeppens please ducument this LTI tool in [documentation](https://github.com/opencast/opencast/blob/r/10.x/docs/guides/admin/docs/modules/ltimodule.md)

kristofkeppens commented:
@wsmirnow forgot about that one, fixed now.

LukasKalbertodt committed:
Add DB index for `modified_date` on `SeriesEntity`

The new `getAllForAdministrativeRead` method performs these date range
queries that would otherwise be very slow.

lkiesow commented on docs/guides/admin/docs/modules/ltimodule.md:
```diff
@@ -108,6 +108,13 @@ custom parameters to be defined globally.
     - `series=SERIESID` if you have the series ID
     - `series_name=SERIESNAME` if you just have the series name (has to be unique)
     - `lng=LANG` to force a language (the browser language is used otherwise)
+- To show a selection dialog to insert opencast content into the LMS via the LTI deeplink protocol
+  ( http://www.imsglobal.org/specs/lticiv1p0 ), use the following parameters
```
```suggestion
- To show a selection dialog to insert opencast content into the LMS via the
  [LTI deeplink protocol](https://www.imsglobal.org/specs/lticiv1p0), use the following parameters
```

kristofkeppens replied:
@lkiesow regarding the unsafe part, is this referring to the LTI deeplinking protocol, or the implementation we made?

lkiesow commented on docs/checkstyle/checkstyle-suppressions.xml:
```diff
@@ -30,6 +30,7 @@
     <suppress checks="LineLength" files="modules/inspection-service-ffmpeg/.+\.(java|properties)"/>
     <suppress checks="LineLength" files="modules/kernel/.+\.(java|properties)"/>
     <suppress checks="LineLength" files="modules/live-schedule-impl/.+\.(java|properties)"/>
+    <suppress checks="LineLength" files="modules/lti/.+\.(java|properties)"/>
```
We shouldn't add exceptions for new features. They should adhere to the rules.

lkiesow commented on modules/lti/package.json:
```diff
@@ -9,6 +9,9 @@
     "bootstrap": "^5.0.1",
+    "dangerously-set-html-content": "^1.0.8",
+    "engage-ui": "../engage-ui",
+    "http-mock-server": "^0.5.2",
```
It should suffice to have this as a dev dependency down below

kristofkeppens replied:
removed engage-ui, moved http-mock-server to dev dependency. Still having a look at the dangerously-set-html-content on what to do with this.

lkiesow commented on modules/lti/package.json:
```diff
@@ -9,6 +9,9 @@
     "axios": "^0.21.1",
     "bootstrap": "^5.0.1",
+    "dangerously-set-html-content": "^1.0.8",
+    "engage-ui": "../engage-ui",
```
Why do we depend on the engage ui?
Given its state, that doesn't feel healthy for a new module.

kristofkeppens replied:
This was to import the css file and reuse this on this page, but I agree that this seems unnecessary, especially given the limited use of that css file, removed the dependency and will include the necessary css directly in the relevant css file for this page.

lkiesow commented on modules/lti/package.json:
```diff
@@ -9,6 +9,9 @@
     "@fortawesome/react-fontawesome": "^0.1.14",
     "axios": "^0.21.1",
     "bootstrap": "^5.0.1",
+    "dangerously-set-html-content": "^1.0.8",
```
> React uses dangerouslySetInnerHtml prop to render raw html, and works pretty much well for almost all the cases, but what if your html has some scripts tags inside??

That sounds… dangerous and somewhat unsafe. Is this well thought through?

karendolan replied:
@kristofkeppens I found a potential option for avoiding that lib, however, it negates the easy way to view the content item data before posting it. It's replacing the dangerous...'s `InnerHTML` with the following `LtiLaunchForm`. Inspired by https://github.com/remarkablemark/html-react-parser/issues/94

```
import React, {useEffect} from "react";
import DOMPurify from "dompurify"; // 2mill+ downloads per week, well supported
import parse from "html-react-parser"; // 600k downloads per week, active

// Input properties used for LTI DeepLinking launch form
interface LtiLauchFormProps {
    // LTI Launch form in a serialized HTML <pre> snippet
    // sent by the tool provider (Opencast)
    // in response to the /lti/ci POST of the video selection.
    readonly formString: string,
}

// The LTI Launch form, returned by Opencast (OC) on /lti/ci POST
// is rendered and submitted to the tool consumer by this component.
// The form submission embeds the video as an LTI Deep Link (DL) launch
// in the host.
// The benefit of a video DL launch is that current host user and course
// context LTI data from the host is passed to Opencast at the time
// the video is launched from the host in the future.
export const LtiLaunchForm: React.FunctionComponent<LtiLauchFormProps> = ({
    formString,
}) => {

    // Auto-Submit the rendered LTI Deep Link form as soon as it's mounted.
    // The user has already selected this video to embed, per the /lti/ci POST.
    useEffect(()=>{
       const forms = document.getElementsByName('ltiLaunchForm');
       // NOTE: The debug/testing javascript from the HTML snippet
       // is scrubbed. It is not possible to view the html prior to posting
       // using this component. To view form post content, use the
       // native browser's inspector/developer tool's network tab
       if (forms.length > 0 && forms[0] instanceof HTMLFormElement) {
           // TODO: error handling for rejected submit?
           // Examples of rejection are when the host does not support
           // LTI DeepLink components or the user's session
           // with the host has timed out.
           forms[0].submit();
       }
    })

    // De-serialize the serialized form snippet
    const cleanString = DOMPurify.sanitize(formString);
    const cleanForm = parse(cleanString);

    // Return the rendered form.
    return (
        <div>
            {cleanForm}
        </div>
)}
```

lkiesow commented on modules/lti/src/OpencastRest.ts:
```diff
@@ -208,21 +215,23 @@ export async function searchEpisode(
-                        : [result.mediapackage.creators.creator]
-                    : [],
+            mediapackage: result.mediapackage !== undefined ? {
+                creators: result.mediapackage.creators !== undefined ? result.mediapackage.creators.creator : [],
```
Why reformat this (same below)?
Readability before was far better.

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiServlet.java:
```diff
@@ -48,12 +56,15 @@
  * produce JSON containing the LTI parameters passed during LTI launch.
  */
-public class LtiServlet extends HttpServlet {
+public class LtiServlet extends HttpServlet implements ManagedService {
```
Why implement a `ManagedService`?

kristofkeppens replied:
@lkiesow it seems that this is necessary to make the following work : `consumerDetailsService.loadConsumerByConsumerKey(consumerKey);`

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiServlet.java:
```diff
@@ -268,6 +319,31 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
+    boolean test = false;
+    if ("true".equals(req.getParameter(TEST))) {
+      test = true;
+    }
```
```suggestion
    final boolean test = "true".equals(req.getParameter(TEST));
```

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiServlet.java:
```diff
@@ -309,4 +395,13 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
     }
   }
+
+  public void activate(ComponentContext cc) {
```
```suggestion
  public void activate() {
```

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiUtils.java:
```diff
@@ -0,0 +1,336 @@
+      return nextProp;
+    } catch (net.oauth.OAuthException e) {
+      logger.warn("BasicLTIUtil.signProperties OAuth Exception {}", e.getMessage());
+      throw new Error(e);
```
Any particular reason to turn the Exceptions into `Error`?
Will the errors be logged?

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiUtils.java:
```diff
@@ -0,0 +1,336 @@
+    text.append("\" name=\"ltiLaunchForm\" id=\"ltiLaunchForm_" + submitUuid + "\" method=\"post\" ");
+    text.append(" encType=\"application/x-www-form-urlencoded\" accept-charset=\"utf-8\">\n");
+    if (debug) {
+    }
```
```suggestion
```

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiUtils.java:
```diff
@@ -0,0 +1,336 @@
+    text.append("   ourUrl = window.location.href;\n");
+    text.append("   if ( formAction.indexOf('http://') == 0 && ourUrl.indexOf('https://') == 0 ) {\n");
+    text.append("      theform.target = '_blank';\n");
+    text.append("      window.console && console.log('Launching http from https in new window!');\n");
```
```suggestion
    text.append("      window.console && console.debug('Launching http from https in new window!');\n");
```

lkiesow commented on modules/lti/src/main/java/org/opencastproject/lti/LtiUtils.java:
```diff
@@ -0,0 +1,336 @@
+    StringBuilder text = new StringBuilder();
+    // paint form
+    String submitUuid = UUID.randomUUID().toString().replace("-","_");
+    text.append("<div id=\"ltiLaunchFormArea_");
```
This whole block is hard to read and seems like it should be a template. I'm also not sure if and how user input is passed in here and if this is safe. E.g. thinking of passing a backspace character in to manipulate the web page.

lkiesow reviewed (changes requested):
This is just a quick review of the code surface, since I don't really have any idea what LTI deep linking is supposed to do. Some parts of this feel somewhat unsafe… but then I'm not sure what to expect… Not sure how to go on with the review.

LukasKalbertodt committed:
Fix and improve error logging in `SeriesServiceDatabaseImpl`

Lots of these were copy & paste bugs. Also, the logger supports
exceptions as last parameter natively. Printing them via `{}` and
`e.getMessage()` is worse.

KatrinIhler committed:
Add option 3 to upgrade guide

KatrinIhler committed:
Remove trailing spaces

KatrinIhler committed:
Differentiate between index identifier and name

KatrinIhler committed:
Delete unused classes

KatrinIhler committed:
Minor fixes

KatrinIhler committed:
Correct docs

KatrinIhler committed:
Fix config parsing bug

KatrinIhler committed:
Harmonize logging

KatrinIhler committed:
Fix trailing spaces AGAIN

KatrinIhler committed:
Add endpoint to resume index rebuild

KatrinIhler committed:
Use basic auth in request example in docs

KatrinIhler committed:
Fix migration scripts

gregorydlogan commented:
Per discussion in the technical meeting today, we're going to close this for now.  When you're ready please feel free to refile this!

kristofkeppens committed:
fix tab series not appearing

kristofkeppens committed:
fixes the series tab not appearing

LukasKalbertodt committed:
Throw `UnauthorizedException` directly from `getAllForAdministrativeRead`

That way it is easier to check the cause of the error instead of having
to unwrap two layers of `SeriesException`s.

LukasKalbertodt committed:
Add SQL migration script for changes to `oc_series`

`deletion_date` and `modified_date` were added. This commit also adds
the corresponding migration script for PostgreSQL since the existing 
script only works/worked for MySQL and MariaDB.

LukasKalbertodt committed:
Add note about DB migration scripts required during upgrade

geichelberger committed:
update osgi compendium and core

KatrinIhler committed:
Elasticsearch index retry configuration

Retry get, update or delete requests if the index is temporarily
unavailable. Number of retry attempts and waiting period between retries
is configurable separetely for read requests and write requests.

KatrinIhler committed:
Small fix

LukasKalbertodt committed:
Add proper `DublinCoreCatalog` (de)serialization when using Gson

The default Gson serialization with this class is rather bad and the
deserialization actually fails. Instead, we now use the already existing
methods for JSON (de)serialization.

LukasKalbertodt committed:
Store deletion and modified date for series in the DB

This is useful for several things, but in particular: we need this for
the upcoming Tobira harvesting API. Without information about when
something about a series changes, a syncing application would have to
request all series every time, which is quite a waste.

To also be able to notify other applications about deleted series, we
cannot delete them for real, but just flag them as deleted. Instead of
a flag, a timestamp is stored as that might be useful in the future.

LukasKalbertodt committed:
Add `SeriesService.getAllForAdministrativeRead`

This method will be used in the Tobira harvesting API to retrieve all
series in a specific time range. Since the dublin core catalogue is not
sufficient for several use cases, a new class `Series` is added as
well.

LukasKalbertodt committed:
Only return series belonging to the current organization in getAllForAdministrativeRead

LukasKalbertodt committed:
Properly implement `getAllForAdministrativeRead` for series remote impl

This was left out from the initial commit adding this method. It uses
the JSON endpoint added last commit.

KatrinIhler committed:
Make upgrade guide clearer

LukasKalbertodt committed:
Add series JSON endpoint `allInRangeAdministrative.json`

This endpoints only purpose is to implement `getAllForAdministrativeRead`
for the remote impl. It is an internal API and should not be used for
any other purpose.

geichelberger committed:
fix remote component typo name

KatrinIhler committed:
Clarify noErrorStateServices config

KatrinIhler committed:
One Elasticsearch index to rule them all (#2857)

mtneug committed:
Allow empty track duration (#3148)

gregorydlogan committed:
Merge branch 'error-state-config' of KatrinIhler/opencast into develop

Pull request #3126
  Add config for service error states to service registry

gregorydlogan committed:
Merge branch '11.x-wowza-str-sec-rfh' of mliradelc/opencast into develop

Pull request #3090
  Wowza stream security  "Prefix:Secret" configuration

gregorydlogan committed:
Merge branch 'wf-src' of lkiesow/opencast into develop

Pull request #3129
  Extract image from source

gregorydlogan committed:
Merge branch 'downgrade-to-latest-xml-apis-version' of geichelberger/opencast into develop

Pull request #3132
  Downgrade to latest xml-apis version from 2.0.2 to 1.4.01

gregorydlogan committed:
Merge branch 'remove-mina' of lkiesow/opencast into develop

Pull request #3134
  Remove Apache Mina

gregorydlogan committed:
Merge branch 'rest-assured-4.4.0' of lkiesow/opencast into develop

Pull request #3143
  Bump rest-assured to 4.4.0

gregorydlogan committed:
Merge branch 'remove-nekohtml' of lkiesow/opencast into develop

Pull request #3144
  Remove com.springsource.org.cyberneko.html

gregorydlogan committed:
Merge branch 'fix-file-ext' of mtneug/opencast into develop

Pull request #3149
  Use original file extension in the asset manger

gregorydlogan committed:
Merge branch 'resume-index-rebuild' of git://github.com/KatrinIhler/opencast into develop

gregorydlogan committed:
Merge branch 'remove-series-title-index-update' of git://github.com/KatrinIhler/opencast into develop

gregorydlogan committed:
Merge branch 'update-osgi' of geichelberger/opencast into develop

Pull request #3131
  Update osgi compendium and osgi core from 5.0.0 to 6.0.0

ts23 committed:
asset-manager move to S3 duplicates identical files #3164

ts23 committed:
asset-manager move to S3 duplicates identical files #3164

fix comment

geichelberger committed:
fix content-type header of ingest endpoint

Fixes issue #3146

geichelberger committed:
add spa style redirect if resource does not exist

The implementation mimics the nginx try_files behaviour. If a resource
does not exist, it returns the welcome page.

Fixes #1494

dependabot[bot] committed:
Bump http-errors from 1.8.0 to 1.8.1 in /modules/engage-paella-player

Bumps [http-errors](https://github.com/jshttp/http-errors) from 1.8.0 to 1.8.1.
- [Release notes](https://github.com/jshttp/http-errors/releases)
- [Changelog](https://github.com/jshttp/http-errors/blob/master/HISTORY.md)
- [Commits](https://github.com/jshttp/http-errors/compare/1.8.0...1.8.1)

---
updated-dependencies:
- dependency-name: http-errors
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump html-validate from 6.1.1 to 6.1.2 in /modules/admin-ui-frontend

Bumps [html-validate](https://gitlab.com/html-validate/html-validate) from 6.1.1 to 6.1.2.
- [Release notes](https://gitlab.com/html-validate/html-validate/tags)
- [Changelog](https://gitlab.com/html-validate/html-validate/blob/master/CHANGELOG.md)
- [Commits](https://gitlab.com/html-validate/html-validate/compare/v6.1.1...v6.1.2)

---
updated-dependencies:
- dependency-name: html-validate
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

KatrinIhler committed:
Save button for series metadata

mtneug committed:
Always update groups (#3157)

There are more then AAI user reference provider that need group updates.
This remove the restriciton for AAI.

rute-santos committed:
New  woh: select-version (#2596)

* New select-version woh, which replaces the current mp with another version from the asset manager.

* Address reviewer comments: better check of operation config, add how to find mp version to documentation.

* Removed part of doc because it isn't reliable.

* Rebased and now using osgi annotations.

* Rebased and fixed conflict.

gregorydlogan committed:
Merge branch 'r/10.x' into develop

dependabot[bot] committed:
Bump bower from 1.8.12 to 1.8.13 in /modules/admin-ui-frontend

Bumps [bower](https://github.com/bower/bower) from 1.8.12 to 1.8.13.
- [Release notes](https://github.com/bower/bower/releases)
- [Changelog](https://github.com/bower/bower/blob/master/CHANGELOG.md)
- [Commits](https://github.com/bower/bower/compare/v1.8.12...v1.8.13)

---
updated-dependencies:
- dependency-name: bower
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

mtneug committed:
Merge pull request #3168 from geichelberger/fix-content-type-header-of-ingest

Fix content-type header of ingest endpoint

gregorydlogan committed:
Merge branch 'allow-asset-track-upload' of mtneug/opencast into develop

Pull request #3150
  Allow asset upload of tracks

gregorydlogan committed:
Merge branch 'execute-logs' of mtneug/opencast into develop

Pull request #3153
  Always ouput execute-* process logs

gregorydlogan committed:
Merge branch 'fix-mimetype-parsing' of mtneug/opencast into develop

Pull request #3158
  Fix mime type parsing for publish-oaipmh operation

gregorydlogan committed:
Merge branch 'remote-service-typo' of geichelberger/opencast into develop

Pull request #3166
  fix remote component typo name

gregorydlogan committed:
Merge branch 'add-spa-routing-support' of geichelberger/opencast into develop

Pull request #3169
  Add spa style redirect if resource does not exist

KatrinIhler committed:
Save button for event metadata

mtneug committed:
Add release notes for tenant URL mapping

gregorydlogan committed:
Merge branch 'tenant-hostmapping' of mtneug/opencast into develop

Pull request #2559
  Allow proper mapping of tenant hostnames to URLs

dependabot[bot] committed:
Bump karma from 6.3.8 to 6.3.9 in /modules/admin-ui-frontend

Bumps [karma](https://github.com/karma-runner/karma) from 6.3.8 to 6.3.9.
- [Release notes](https://github.com/karma-runner/karma/releases)
- [Changelog](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md)
- [Commits](https://github.com/karma-runner/karma/compare/v6.3.8...v6.3.9)

---
updated-dependencies:
- dependency-name: karma
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

lkiesow committed:
Merge r/10.x into develop

KatrinIhler committed:
Fixes

ts23 committed:
asset-manager move to S3 duplicates identical files #3164

fix job-producer tests

KatrinIhler committed:
Merge pull request #3159 from mtneug/rm-track-check

Remove check for tracks for OAI-PMH

gregorydlogan committed:
Merge branch 'fix#3164_assetmanager_move_to_S3' of ts23/opencast into develop

Pull request #3167
  Fix#3164 assetmanager move to s3

gregorydlogan committed:
Merge branch 'metadata-save-buttons' of KatrinIhler/opencast into develop

Pull request #3176
  Save Buttons for Metadata

mtneug committed:
Remove automatic handling of HLS bitrate ladder (#2591)

* Remove automatic handling of HLS bitrate ladder

When using HLS Opencast currently forces the use of `-b:v` and `-b:a`
for each stream. The bitrate is determined either from the same options
in the encoding profile or geometrically distribute between a min and a
max. This limits the rate control mechanism to either Average Bitrate
(ABR) (i.e. only setting `-b`) or Constant Bitrate (CBR) (i.e. setting
`-b`, `-maxrate`, `-minrate`, `-bufsize`). In particular, Constrained
Encoding (VBV) (i.e. setting `-crf`, `-maxrate`, `-minrate`, `-bufsize`)
is not possible.

This removes the bitrate handling in Opencast completely forcing the
user to properly setup the rate control in the used encoding profiles.

* Remove mentions of removed HLS properties in documentation

* Add HLS bitrate handling release notes

Gao-Jun committed:
Encode username before useing it in CanvasUserRoleProvider (#2949)

Arnei committed:
Removed Solr from Series Service (#3161)

* Removed Solr from Series Service

Removes all Solr functionality from the series service. Functions that
relied on the solr index were either removed or now query the database
directly.

- Moved the series filter query to the external api, while trying to
  preserve as much functionality as possible (for example, there is no
pagination, so instead an offset variable is offered)

* Fixed wrong variable name in SeriesEndpoint.java

* Fixed missing imports

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Update osgi compendium in 10.x merge

This applies the changes from pull request #3131 to the merges from
r/10.x to 11.x.

lkiesow committed:
Update Node to Latest LTS

This patch updates Node to version 16.13.0 which is currently the latest
LTS version of Node.

geichelberger committed:
use jdk xalan version

geichelberger committed:
update batik from 1.7 to 1.14

geichelberger committed:
remove deprecated methods and remove duplicate schema url

Commit:
Automatically update translation keys (r/11.x)

KatrinIhler committed:
Allow index name to be changed at runtime

...since it's only used for logging.

Arnei committed:
Fixed Admin UI Endpoint Configuration not loading on Opencast startup

As mentioned in PR #3049 by @KatrinIhler, said PR caused some of the admin ui endpoint configuration files to not be read on Opencast start-up anymore, but only when the configuration files were modified at runtime. This fixed that.

Arnei committed:
Just get Map as a parameter, instead of parsing it from a dictionary

KatrinIhler committed:
Merge pull request #3198 from Arnei/config-not-read-in-admin-ui

Fixed Admin UI Endpoint Configuration not loading on Opencast startup

KatrinIhler committed:
Merge pull request #3127 from lkiesow/adaptive-playlist

Fix minor bugs in AdaptivePlaylist

gregorydlogan committed:
Merge branch 'node-16.13.0' of git://github.com/lkiesow/opencast into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

mliradelc committed:
Revert "Removed Solr from Series Service (#3161)" (#3201)

This reverts commit 006accc3f8120de7cbe60b6200eadb0352227013.

jduehring committed:
Replace native javascript with angularJS code (#3177)

This patch replaces some inline JavaScript code in the admin interface,
which should not have made it into Opencast.
The code is responsible for displaying the embed code for each episode.

This fixes #2786

Co-authored-by: Jonas Duehring <duehring@elan-ev.de>
Co-authored-by: Lars Kiesow <lkiesow@uos.de>

lkiesow committed:
Merge r/10.x into r/11.x

ts23 committed:
Update Start Workflow WFH to take comma separated list of mediapackages
Update Duplicate Workflow WFH to provide comma separated list of mediapackages
Modify the Duplicate Event Workflow to demonstrate Duplicate WFH and StartWorkflow WFH in combination

# Conflicts:
#	docs/guides/admin/docs/workflowoperationhandlers/start-workflow-woh.md
#	etc/workflows/duplicate-event.xml
#	modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java

lkiesow committed:
Improve GitHub Actions Concurrency Configuration

This patch improves the GitHub Actions concurrency configuration by
moving the documentation deployment into it's own concurrency namespace
and by ensuring that the last commit on a branch will win when deploying
to Crowdin.

lkiesow committed:
Link Configuration Files in GitHub

This patch automatically links configuration files referenced in the
documentation to the actual configuration files in GitHub. This is done
in addition to the tooltip added to `etc/` explaining where to find
these files.

lkiesow committed:
Merge r/10.x into r/11.x

Commit:
Automatically update translation keys (r/11.x)

lkiesow committed:
Extend Metadata Extraction in Inbox

Opencast's inbox allows you to add zipped media packages which may
contain arbitrary metadata but are very complex or to add simple files,
which file names will end up as the new event's title. It would be nice
if it would be possible to extract a few more basic metadata from the
filename.

File names like these are commonly generated by capture devices:

```
2020-02-07-hs1-Introduction to Biology.mp4
 |    |  |  |  |                     |
year  |  |  |  +----------+----------+
   month |  |             |
        day |           title
        location
```

To improve the situation, and to be as flexible as necessary, we would
like to allow for specifying regular expressions as matchers for a few
basic metadata.

This resolves #3239

lkiesow committed:
Prevent API failing on empty files

This patch fixes the problem that the REST docs include fields to upload
files foor the `POST /api/events` requests which will result in empty
fields being send when trying to schedule over which Opencast will
stumble and fall (500 Internal Server Error).

This patch makes the API ignore empty file fields in the
multipart/form-data request.

This fixes #3249

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

mtneug committed:
Fix posible type change in external API (#3231)

PR #3148 introduced a possible type change in the external API from int
to string for the duration field of tracks in case the field is `null`.
As discussed, this PR changes the return value to JSON `null` instead.

Note that the behavior is still inconsistend when it comes to durations:

* `{eventId}/media` will omit the field
* Event duration has a type change to `""`
* this changes Track duration to `null`

This patch leave the behavior of `{eventId}/media` as this is done with
all the fields. The Event duration though now also returns JSON `null`
as default value.

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Switch to GitHub icon for documentation link

This patch switches to the actual GitHub icon for links to configuration
file in the git repository on docs.opencast.org.

gregorydlogan committed:
Merge pull request #3221 from lkiesow/link-configuration-files

Link Configuration Files in GitHub

gregorydlogan committed:
Merge pull request #3220 from lkiesow/workflow-concurrency

Improve GitHub Actions Concurrency Configuration

mtneug committed:
Add documentation for using organisation properties in workflows

mtneug committed:
Add documentation for id and org_id in the execute service

mtneug committed:
Add documentation for tag selection in execute service

mtneug committed:
Merge pull request #3128 from lkiesow/encode-hls

Let Encode Handle HLS

mliradelc committed:
Merge branch 'r/10.x' into r/11.x

lkiesow committed:
Update Logging Documentation

This patch updates the logging documentation, fixing a number of syntax
errors and broken links.

Commit:
Automatically update translation keys (r/11.x)

Commit:
Remove chinese traditional (#3263)

Since for chines traditional only 68% of the
strings are translated, it will be removed from
OC 11.

Commit:
Prepare docs for OC 11 release (#3265)

* Prepare docs for OC 11 release

* Update changelog

Commit:
Fix formatting issues in release notes (#3266)

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Fix Typos and Syntax in Opencast 11 Release Notes

This patch fixes some typos and minor syntax problems in the Opencast 11 release notes.

lkiesow committed:
Remove old release notes snippets

The release notes for Opencast 11 have been written and the snippets
should have been cleaned out to not mix them up with the notes for the
next release.

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Adding release notes for OC 11.1

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Mark Opencast 11 as Available

This patch marks the Opencast 11 RPMs as being available.

narickmann committed:
Changed embed code selection design (#3273)

This fixes #3214

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Add missing 11.2 changelog

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Java Dependency Update

This patch updates several Java dependencies to their latest versions.

LukasKalbertodt committed:
Use 1970-01-02 as dummy value in MySQL migration

This is a workaround for timezone issues. The exact value doesn't matter, so just 
adding a day is the easiest option to fix problems like this:
https://groups.google.com/a/opencast.org/g/users/c/335SnoOC2tQ

Commit:
Removed ManagedService from MediaInspectionService

This hopefully fixes #3217
Not sure, if this is correct (first time  OSGi).

lkiesow committed:
Remove non-optional optional

This removes the use of some of Opencast's own Optional classes where
the value is actually non-optional which actually makes the code simpler
as well.

lkiesow committed:
Match Inbox Events Against Schedule

Capture agents often record to filenames containing metadata which
allows us to identify an event by looking at a specific time in the
schedule of a certain location (capture agent).

If `created` and `spatial` metadata could be identified, this patch
allows us to look up the schedule, using all available event data fram
there if we had a unique match.

Data taken from the schedule, overwriting inbox data are:

- All metadata
- Workflow definition
- Workflow configuration

Notes:

- This patch fixes #3240.
- This is based on #3327.
- You want #3339 for this.

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Fix REST Test Netwprk Bindings

Using the REST tests environment, Opencast's tests would randomly fail
with ports seemingly being already in use, causing the bind of the test
server to fail.

This patch changes the test environment behavior to select a random port
and the directly trying a bind instead of selecting a port in advance.
This should cause the test server to select another port if the bind
fails.

This (hopefully) fixes #2098

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

lkiesow committed:
Use ffprobe to get additional metadata

In addition to the metadata extraction from the file name, this patch
allows users to have ffprobe run on the media files to get some
additional metadata from the media container.

The extracted metadata are creation time and duration, which may help to
match scheduled events.

With the media file now having a duration, there is now also an
additional option allowing to filter out some events which are barely
conflicting (imaging two back to back recordings and the second one
starts a minute early).

lkiesow committed:
Allow metadata editting

This patch removes a check which would prevent users from editing
metadata of scheduled events which start time has already passed.

Editing such metadata in the admin interface will fail silently with no
feedback for users but many errors in the logs:

```
2022-01-10 21:42:32,986 | INFO  | (SchedulerServiceImpl:837) - Event ID ba78ec72-6d8c-40f8-807c-67d6a6764a39 has already ended as its end time was 2022-01-10T16:10:00Z and current time is 2022-01-10T20:42:32Z
2022-01-10 21:42:32,987 | ERROR | (IndexServiceImpl:1303) - Unable to update scheduled event ba78ec72-6d8c-40f8-807c-67d6a6764a39 with metadata [...] because
org.opencastproject.scheduler.api.SchedulerException: Event ID ba78ec72-6d8c-40f8-807c-67d6a6764a39 has already ended at 2022-01-10T16:10:00Z and now is 2022-01-10T20:42:32Z
        at org.opencastproject.scheduler.impl.SchedulerServiceImpl.verifyActive(SchedulerServiceImpl.java:840) ~[?:?]
        at org.opencastproject.scheduler.impl.SchedulerServiceImpl.updateEventInternal(SchedulerServiceImpl.java:671) ~[?:?]
        at org.opencastproject.scheduler.impl.SchedulerServiceImpl.updateEvent(SchedulerServiceImpl.java:619) ~[?:?]
```

I see no reason why we shouldn't allow users to edit e.g. the title of
such events and this patch now makes it possible.

gregorydlogan committed:
Merge branch 'cover-image-surrogate-problem' of geichelberger/opencast into r/11.x

Pull request #3188
  Cover image surrogate problem

lkiesow committed:
Merge r/10.x into r/11.x

gregorydlogan committed:
Merge branch 'java-dependency-updates' of lkiesow/opencast into r/11.x

Pull request #3328
  Java Dependency Update

gregorydlogan committed:
Merge branch 'rest-test-port' of lkiesow/opencast into r/11.x

Pull request #3336
  Fix REST Test Network Bindings

gregorydlogan committed:
Merge branch 'non-optional' of lkiesow/opencast into r/11.x

Pull request #3337
  Remove non-optional optional

gregorydlogan committed:
Merge branch 'api-ignore-empty-files' of lkiesow/opencast into r/11.x

Pull request #3254
  Prevent API failing on empty files

gregorydlogan committed:
Merge branch 'add-id-org_id' of mtneug/opencast into r/11.x

Pull request #3152
  Add id to execute-many and org_id to execute-many and execute-once

gregorydlogan committed:
Merge branch 'add-tags-execute-once' of github.com:mtneug/opencast into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

marwyg committed:
Added speech to text (Vosk)

This commit adds a new workflow operation that can generate subtitles with Vosk.

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Remove `mvn site` Test

This patch removes the `mvn site` test from `r/11.x` and `develop` since
it will currently let pull requests fail. We can easily re-add this
together with the Javadoc fixes for `r/11.x`

Commit:
Add release notes and changelog for OC 11.3 (#3367)

lkiesow committed:
Capture agent calendar as JSON

This patch provides the capture agent calendar in JSON format. This
allows agents to talk to Opencast and get all necessary responses as
JSON which makes handling these data much simpler.

The new endpoint `/recordings/calendar.json` returns data like this:

```json
[
  {
    "eventId": "95b7b721-5d42-4073-9222-634e5cfd8aaa",
    "agentId": "a",
    "startDate": "Jan 19, 2022, 11:30:00 PM",
    "endDate": "Jan 19, 2022, 11:55:00 PM",
    "presenters": [],
    "workflowProperties": {
      "straightToPublishing": "true"
    },
    "agentConfig": {
      "capture.device.names": "",
      "org.opencastproject.workflow.definition": "fast",
      "org.opencastproject.workflow.config.straightToPublishing": "true",
      "event.location": "a",
      "event.title": "test"
    },
    "recording": {}
  }
]
```

Which is a serialized list of the scheduler's TechnicalMetadata class.

lkiesow committed:
Update slf4j

This patch updates slf4j to a fairly recent version. This fixes a few
security issues which aren't really relevant since they are used only in
the tests. Nevertheless, it is nice to not have these issues in the
code at all.

lkiesow committed:
Update modules/scheduler-impl/src/main/java/org/opencastproject/scheduler/endpoint/SchedulerRestService.java

gregorydlogan committed:
Merge branch 'slf4j' of lkiesow/opencast into r/11.x

Pull request #3369
  Update slf4j

gregorydlogan committed:
Merge branch 'scheduler-json' of lkiesow/opencast into r/11.x

Pull request #3368
  Capture agent calendar as JSON

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Merge branch 'edit-scheduled-data' of lkiesow/opencast into r/11.x

Pull request #3342
  Allow metadata editting

gregorydlogan committed:
Merge branch 'add-org-prop-wf' of mtneug/opencast into r/11.x

Pull request #3155
  Allow the use of organization properties in workflows

gregorydlogan committed:
Merge branch 'inbox-metadata' of lkiesow/opencast into r/11.x

Pull request #3327
  Extend Metadata Extraction in Inbox

lkiesow committed:
Add DublinCore Catalog to Scheduler JSON

This patch adds the episode DublinCore catalog to the JSON version of
the capture agent calendars.

The new format is:

```json
[
  {
    "agent": {…},
    "episode-dublincore": {…}
  }
]
```

ts23 committed:
Update Start Workflow WFH to allow media-package and media-packages

# Conflicts:
#	etc/workflows/duplicate-event.xml

ts23 committed:
Update Start Workflow WFH to consider fail-on-error for fast fail or collect errors

# Conflicts:
#	modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/StartWorkflowWorkflowOperationHandler.java

lkiesow committed:
Merge r/10.x into r/11.x

ts23 committed:
Update docs/guides/admin/docs/workflowoperationhandlers/start-workflow-woh.md

Co-authored-by: Matthias Neugebauer <mtneug@mailbox.org>

ts23 committed:
Update docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md

Co-authored-by: Matthias Neugebauer <mtneug@mailbox.org>

mtneug committed:
Merge pull request #3203 from ts23/start-workflow-wfh_add_media-package-id-key

start-workflow-wfh allows to pass mediapackage id from previous WFH

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Fixed Typo in Logging Statement

This patch fixes a typo in the logging statement which was complaining
about the wrong group not being matched.

geichelberger committed:
add DownloadDistributionService unbind method

lkiesow committed:
Merge r/10.x into r/11.x

gregorydlogan committed:
Merge branch 'inbox-schedule' of lkiesow/opencast into r/11.x

Pull request #3340
  Match Inbox Events Against Schedule

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Merge r/10.x into r/11.x

mtneug committed:
Merge pull request #3384 from lkiesow/scheduler-json-dc

Add DublinCore Catalog to Scheduler JSON

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Fix `pom.xml` Version

This patch fixes the `pom.xml` version of the rename file workflow
operation handler.

lkiesow committed:
Fix merge conflict

This patch fixes the problem of the OSGi compendium having been renamed
in newer versions. Since the rename-files operation was merged from
Opencast 10 and still used the old name, this caused a build failure.

kristofkeppens committed:
Merge the ldap userdirectory behaviour from 9.x into the current module. 

This makes the module work as it did in 9.x when using it only for authorization. This does not implement all the changes present in the OpencastLdapAuthoritiesPopulator (#3344)

lkiesow committed:
Fix frontend-no-prebuilt profile (#3444)

This patch fixes the frontend-no-prebuilt maven profile which can be
used to have maven run the local npm version instead of downloading one.
That has the big advantage of not having to remove any `none*` folders
when you switch versions.

To try this, run:

```
❯ mvn clean install -Pdev,frontend-no-prebuilt
```

The profile did exist for a long time but was broken due to some tasks
being executed in the wrong order like trying to copy the frontend to
the target folder before actually building it.

LukasKalbertodt committed:
Updates to Opencast Studio (#3457)

https://github.com/elan-ev/opencast-studio/releases/tag/2022-02-16

* Added Chinese simplified translations (thanks to Phyo Pa!)
* Updated dependencies

Commit:
Add release notes and changelog for OC 11.4 (#3459)

* Add release notes and changelog for OC 11.4

* Fix linting error

darolfes committed:
fix admin-ui endpoint crash, when a series without title exists

CGreweling committed:
Check if encoding profile exist for imageToVideo Operation

Update modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

darolfes committed:
fix admin-ui endpoint crash, when a series without title exists: Apply suggestions from code review

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

Commit:
Fix OC 11.4 release notes layout issues

gregorydlogan committed:
Merge branch 'empty-series-bug-fix-r11.x' of darolfes/opencast into r/11.x

Pull request #3460
  fix admin-ui endpoint crash, when a series without title exists

lkiesow committed:
Merge r/10.x into r/11.x

narickmann committed:
Password strength indicator (#3343)

This fixes #3215.

Simple strength bar, checks if the password contains at least one uppercase letter, one lowercase letter, one number, one special character and is at least 8 characters long.

kristofkeppens committed:
fix pagination issue for deeplink module

Commit:
Automatically update translation keys (r/11.x)

CGreweling committed:
Force same sample rate for audio files in PartialImport WOH encoding profiles
missing changes from pr#3107

Commit:
Added proposed changes

Hopefully adressed all remaining open points.

kristofkeppens committed:
Rework the userdirectory-brightspace module

Rework the userdirectory module to allow mapping of LMS roles to opencast roles, mimicking the way the LTI roles work so users will get the same role regardless of the way they login

mtneug committed:
Fix wrong name of admin UI configuration file (#3532)

The admin UI configuration file was wrongly named
`org.opencastproject.adminui.cfg` instead of
`org.opencastproject.adminui.impl.AdminUIConfiguration.cfg`

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

lkiesow committed:
Fix Admin UI builds due to missing SASS

This patch fixes the `package.json` which broke when merging `r/10.x` to
`r/11.x` and was missing `sass` as dependency.

mtneug committed:
Merge pull request #3534 from lkiesow/admin-ui-builds-sass

Fix Admin UI builds due to missing SASS

KatrinIhler committed:
Review Fixes

gregorydlogan committed:
Merge branch 'elasticsearch-retry' of KatrinIhler/opencast into r/11.x

Pull request #3197
  Elasticsearch Retry Config

Commit:
Automatically update translation keys (r/11.x)

gregorydlogan committed:
Revert "Remove chinese traditional (#3263)"

This reverts commit f9be02af08dea3ea848d57db808c61a4d8f48cc2.

gregorydlogan committed:
Applying current state of translations from crowdin

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

gregorydlogan committed:
Revert "Revert "Update paella player to paella 6.5.5 (#3142)""

This reverts commit 16725ecbb7bf6466e823e96e73bbb465b9a5ca6b.

gregorydlogan committed:
Merge pull request #3545 from gregorydlogan/t/update-11.x-translations

Update 11.x translations, restore zh_TW translations

jojoob committed:
Downgrade commons-io due to a bug, see #3510 (#3541)

mtneug committed:
Fix NPE when signing publication URLs (#3540)

Publication URLs might be null. Signing them leads to a NPE.

lkiesow committed:
Update Editor to 2022-03-22 (#3546)

This patch updates the stand-alone video editor to release 2022-03-22.
For more details and the full change logs, take a look at the [release
notes](https://github.com/opencast/opencast-editor/releases/tag/2022-03-22).

gregorydlogan committed:
Revert "Revert "Revert "Update paella player to paella 6.5.5 (#3142)""" (#3547)

This reverts commit 4a31a2138956be5de6942812d43c00ced9768219.

darolfes committed:
fix safari sometimes not skipping disabled sections in the editor (#3544)

Commit:
Add release notes and changelog for OC 11.5 (#3550)

* Add release notes and changelog for OC 11.5

* Update release notes

mliradelc committed:
Merge branch 'r/10.x' into r/11.x

mtneug committed:
Fix workflow definition installation

When installing workflow definitions there is a chance that a workflow
definition which fails to load will result in stalling the boot of
further services. This happens when the faulty workflow definition is
loaded last and the `ReadinessIndicator` service is not registered.
This rearranges the code to make sure the service is always registered
when the last workflow definition was handled.

Commit:
Automatically update translation keys (r/11.x)

Alaschinger committed:
Auto-generated series for Capture Agents

Integrated code from switchcast in the ingest-service module to auto-generate a series, if an event without series is ingested. Events without series may be inaccessible for users without admin rights. So this patch provides, that people with rights to the CA have access to the respective series and the contained events.

gregorydlogan committed:
Merge branch 'managed-service' of naweidner/opencast into r/11.x

Pull request #3330
  Removed ManagedService from MediaInspectionService

gregorydlogan committed:
Merge branch 'fix-workflow-def-install' of mtneug/opencast into r/11.x

Pull request #3567
  Fix workflow definition installation

gregorydlogan committed:
Merge branch 'fix/brightspace-userdirectory' of mm-dict/opencast into r/11.x

Pull request #3555
  Rework the userdirectory-brightspace module

JamesUoM committed:
fix #3602, check that host is not in maintenace, due to state from a previous instance

lkiesow committed:
Merge r/10.x into r/11.x

lkiesow committed:
Fix Vosk Concurrency Problem

This patch fixes a concurrency problem with the Vosk based speech to
text engine which might fail due to parallel removal of directories in
the workspace which may make the Vosk command fail.

lkiesow committed:
Error handling in Vosk module

This patch improves the error handling in the Vosk module, actually
providing some feedback to see why the command failed.

It also fixes some minor typos, … along the way.

tyfranck committed:
fix Enable support for automatic punctuation for Google transcription service #2148

kristofkeppens committed:
allow admin user for node to node communication

tyfranck committed:
Fix Enable support for Google Speech-to-text API premium/enhanced service (Video) #2147

gregorydlogan committed:
Merge branch 't/termination-service-persists-maintenance' of UoM-Podcast/opencast into r/11.x

Pull request #3605
  fix #3602, Terminate state service, check that host is not in maintenance

gregorydlogan committed:
Merge branch 'vosk-concurrency' of lkiesow/opencast into r/11.x

Pull request #3630
  Fix Vosk Concurrency Problem

gregorydlogan committed:
Merge branch 'vosk-error-handling' of lkiesow/opencast into r/11.x

Pull request #3631
  Error handling in Vosk module

tyfranck committed:
Remove trailing space reported by configuration test

gregorydlogan committed:
Merge branch '2247_2248_add_punctuation_and_models_support_to_google_transcription' of tyfranck/opencast into r/11.x

Pull request #3638
  Fix add punctuation and models support to google transcription

tyfranck committed:
Fix indentation and document newly added properties following review

gregorydlogan committed:
Merge branch '2247_2248_add_punctuation_and_models_support_to_google_transcription' of tyfranck/opencast into r/11.x

Pull request #3638
  Fix add punctuation and models support to google transcription

lkiesow committed:
Merge r/10.x into r/11.x

gregorydlogan committed:
Merge branch 'r/10.x' into r/11.x

karendolan commented:
Hi @kristofkeppens, my site relies on an early version of your LTI Deeplinking OC implementation. We are upgrading and need to adapt it to the current stable OC. I'm heavily leaning on your pull to help guide the way. If my work is general enough, I will make a pull onto your feature/lti-deeplinking branch.

kristofkeppens commented:
Hi @karendolan , if it helps I can update it to our latest version ( for opencast 11.x ).

kristofkeppens committed:
Merge branch 'r/11.x' into feature/lti-deeplinking

karendolan commented:
> Hi @karendolan , if it helps I can update it to our latest version ( for opencast 11.x ).

Yes! That would be great, thank you @kristofkeppens

kristofkeppens committed:
fix lti deeplink module for 11.x

kristofkeppens committed:
updates to dependencies and fix missing dependency

kristofkeppens commented:
@karendolan I've updated the branch ( https://github.com/mm-dict/opencast/tree/feature/lti-deeplinking ), should be working, if anything isn't working have a look at our https://github.com/mm-dict/opencast/tree/r/11.x-ugent branch, but this contains some other things as well.

karendolan commented:
@lkiesow @kristofkeppens was the LTI content item feature abandoned in upstream Opencast? We have been using this feature for at least 6 years. Our version may be slightly custom. Is it worth creating a new pull?

karendolan commented:
Will use this issue ticket for working on fresh pull

karendolan commented:
Opps, will open a new pull with issue ticket #2164
